### ISSUE TITLE
Unify documentation algebra with core algebra

### ISSUE THREAD

**Issue by Fristi** [julienrf/endpoints#64]
Splitting from #57

I think we can add document parameters to the core algebra (with default values) such that we can drop the documentation algebra. Extending one algebra is easier to maintain. Now if I would like to add for example a `BigDecimal` path segment or any other exotic type I need to change the documentation algebra and the core algebra.

**Comment by Krever** [comment]
That was my original thought as well in #49. Pasting my question and most relevant part of @julienrf answer for reference:

> Would it be possible, to have some "dummy defaults" in Documented API, if someone don't target proper docs? This way we could potentially avoid having two distinct algebras to maintain.


>Now, can we always provide default values? So far, yes: there are only two places where more information is needed: path segments and response descriptions. I just explained which default values we could generate for path segments. For response descriptions that’s even easier: we could just generate “No entity”, “Text entity” and “JSON entity” for the emptyResponse, textResponse and jsonResponse methods, respectively.
>
>What would be the drawback of unifying the DocumentedEndpoints and Endpoints algebras? So far I described the minimal changes to apply to get something work, but in practice OpenAPI can contain a lot more information, and our algebra should make it possible to leverage any feature of OpenAPI. It means, for instance, that we should add parameters to our endpoint method to get a short summary, a long description, whether it is deprecated or not, etc. (all these information can be shown in the OpenAPI description). I am not sure I want to put all these things in the original API.

So I find these arguments quite convincing but I would still love to see unified version. In case we want to keep simple version of methods we could always go with two overrides, one with defaults hardcoded in the body, second one with defaults as optional arguments.

**Comment by Fristi** [comment]
> In practice OpenAPI can contain a lot more information, and our algebra should make it possible to leverage any feature of OpenAPI. It means, for instance, that we should add parameters to our endpoint method to get a short summary, a long description, whether it is deprecated or not, etc. (all these information can be shown in the OpenAPI description). I am not sure I want to put all these things in the original API.

Aha, I see the point :-) We could be opinionated and do a subset of it. Question is are we aiming for OpenApi specification compliance or are we building a library which increase productivity and ensures that documentation is in sync? Personally I would go for the latter as I see already _great_ value in that, but that's my opinion :-)

**Comment by julienrf** [comment]
> Question is are we aiming for OpenApi specification compliance or are we building a library which increase productivity and ensures that documentation is in sync?

Yeah, that’s the key question. My initial attempt was to show that it was possible to provide a “documenting” algebra interface able to integrate with any specific tool like, _for instance_, OpenApi. My guess was that each application or product will have specific needs and that trying to provide one API that would match all different kinds of needs would be worthless.

My vision of an _ideal_ library is something that doesn’t take strong decisions in place of the application developer. That being said, I agree that implementing something that covers 80% of the use cases would still provide a great value.

**Comment by julienrf** [comment]
An update on my feeling with regards to this issue: I found that dealing with the `delegate` interpreters is a bit complicated and indeed having a single set of algebras and interpreters would be simpler. We should really make sure that the interpreter producing documentation always work and produces a meaningful output (even in cases users don’t supply any documentation related information).
